### ISSUE TITLE
Show in sample import dialog if a certain sample is already imported (and don't re-import)

### ISSUE THREAD

**Issue by maks** [democloid/picoTracker#34]

**Comment by democloid** [comment]
@maks There are a number of improvements that could be made regarding samples tracking that could be combined into one properly made feature. Off the top of my head:
1. Tracking imported samples (not by name, probably by hash)
2. Deleting importing samples (this required handling of flash and fragmentation)
3. Tracking sample memory usage
4. Tackling the issue still pending of potential memory leaks or over usage on larger directories

**Comment by maks** [comment]
@democloid agreed on all points! I do have a few thoughts on some of the points:
1. I think hashing could be quite expensive cpu, ram and time wise, especially for longer sample files. Perhaps just filename+filesize would be enough? This could be a problem for single cycle files where they could all be the same length, but for this case, we could just warn users that a sample with the same name already exists in the project and prompt: `overwrite/rename?`
2. yes agreed and on this topic, I have been worried about the current use of the same part of flash for loading a projects samples everytime, especially if only a few small samples are used in the project. So to help with that and fragmentation on delete, it may be worth now looking to use a real filesystem for that part of flash storing current projects samples? There are *lots* of options in this space to choose from, [a few are discussed in this thread](https://www.reddit.com/r/embedded/comments/mvoi23/choosing_an_appropriate_file_system_for_spi_nor/) but there are others too.
3. 💯  on this. I think it would make most sense to show this on the sample import dialog screen.
4. For me this would be the lowest priority as apart from the [bug I need to urgently fix](https://github.com/democloid/picoTracker/issues/83) I think the import works "well enough" for now.

**Comment by maks** [comment]
Just following up on the FS part, https://github.com/littlefs-project/littlefs looks very good, very mature and still very actively maintained.

**Comment by shockdesign** [comment]
My approach was very pragmatic.. however it raises one question.. 

Say a user imports two samples with the same name from two different directories, how are they supposed to know the difference when selecting them?

As for file hash/checksum, maybe we should use something like `CRC32` as it is quite fast ..

```
#include <inttypes.h> // uint32_t, uint8_t

uint32_t CRC32(const uint8_t data[], size_t data_length) {
	uint32_t crc32 = 0xFFFFFFFFu;
	
	for (size_t i = 0; i < data_length; i++) {
		const uint32_t lookupIndex = (crc32 ^ data[i]) & 0xff;
		crc32 = (crc32 >> 8) ^ CRCTable[lookupIndex];  // CRCTable is an array of 256 32-bit constants
	}
	
	// Finalize the CRC-32 value by inverting all the bits
	crc32 ^= 0xFFFFFFFFu;
	return crc32;
}
```

**Comment by maks** [comment]
@shockdesign its not so much the speed of the hashing alogrithm, its that it needs to read through the whole file to calculate it and the speed of that is not great for larger files, even when reading from the onboard flash. But I guess we could do that only once during sample import when we have to read the whole file from the SDCard anyway and then cache it in the project file.

**Comment by shockdesign** [comment]
@maks yeah definitely something we only do once and store... even if we're doing it when we go to check if the file already exists, then just pass it through so we never have to do it again.

**Comment by shockdesign** [comment]
@maks the other thing is we can do it again potentially on project load to verify the samples haven't changed? But thats not as important.
